### PR TITLE
Fix crash when lowering Carbon derived class with C++ virtual base class

### LINKED ISSUE #7721: Crash when lowering a Carbon derived class that overrides a C++ base class with virtual functions
### Description of the bug:

_No response_

### What did you do, or what's a simple way to reproduce the bug?

```carbon
import Cpp inline '''
struct X { virtual void f() {} };
''';

class Z {
    extend base: Cpp.X;
    override fn f(unused self) {}
}

fn Run() {
    var _: Z = {.base = Cpp.X.X()};
}
```

### What did you expect to happen?

_No response_

### What actually happened?

[Crash](https://carbon.godbolt.org/z/4be4W5s17):

```
Stack dump:
[...]
 #4 0x00005fc4256e4c12 Carbon::Lower::FileContext::BuildVtable(Carbon::SemIR::Vtable const&, Carbon::SemIR::SpecificId) (/opt/compiler-explorer/carbon-trunk/bin/carbon+0x248bc12)
 #5 0x00005fc4256e458f Carbon::Lower::FileContext::PrepareToLower() (/opt/compiler-explorer/carbon-trunk/bin/carbon+0x248b58f)
```

### Any other information, logs, or outputs that you want to share?

_No response_

### PR DESCRIPTION
**Pull request by jaswanthsanjay88**
If a Carbon class overrides virtual functions from a C++ base class but is never referenced from C++, it is never exported to Clang. During lowering, `BuildVtable` then fails to find a `CXXRecordDecl` and crashes when attempting to get the vtable from Clang's code generator.

Ensure dynamic classes with foreign vtables are exported to Clang when completing the class definition in `CheckCompleteClassType`, and look up `first_decl_id()` in `BuildVtable`.

Fixes #7721

### CODE REVIEW

**Comment by danakj** [comment]
> Ensure dynamic classes with foreign vtables are exported to Clang when completing the class definition in `CheckCompleteClassType`, and look up `first_decl_id()` in `BuildVtable`. @zygoloid

Tagging people in a PR description adds a lot of github notification noise, so it's best to not do that. I will remove that for you.

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/primitives.carbon
```
Comment (danakj) [inline toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]:
```suggestion
// INCLUDE-FILE: toolchain/testing/testdata/min_prelude/none.carbon
```
Doesn't look like this file needs primitives (i32, bool, etc)

File: toolchain/check/class.cpp [hunk 2]
```diff
@@ -417,6 +417,13 @@ static auto CheckCompleteClassType(
     auto vptr_type_id = GetPointerType(context, SemIR::VtableType::TypeInstId);
     class_info.vtable_decl_id = AddInst<SemIR::VtableDecl>(
         context, node_id, {.type_id = vptr_type_id, .vtable_id = vtable_id});
+    if (!context.vtables().Get(vtable_id).carbon_native_vtable &&
+        class_info.self_type_id.has_value() &&
+        class_info.self_type_id != SemIR::ErrorInst::TypeId) {
+      auto class_type =
+          context.types().GetAs<SemIR::ClassType>(class_info.self_type_id);
+      ExportClassToCpp(context, class_type);
+    }
```
Comment (danakj) [inline toolchain/check/class.cpp]: I don't see how the self_type_id is ever left as None or even set to an ErrorInst in handle_class.cpp. I think this can all be simplified as follows?

```suggestion
    if (!context.vtables().Get(vtable_id).carbon_native_vtable) {
      ExportClassToCpp(context, context.types().GetAs<SemIR::ClassType>(class_info.self_type_id));
    }
```

File: toolchain/check/class.cpp [hunk 3]
```diff
@@ -417,6 +417,13 @@ static auto CheckCompleteClassType(
     auto vptr_type_id = GetPointerType(context, SemIR::VtableType::TypeInstId);
     class_info.vtable_decl_id = AddInst<SemIR::VtableDecl>(
         context, node_id, {.type_id = vptr_type_id, .vtable_id = vtable_id});
+    if (!context.vtables().Get(vtable_id).carbon_native_vtable &&
+        class_info.self_type_id.has_value() &&
+        class_info.self_type_id != SemIR::ErrorInst::TypeId) {
+      auto class_type =
+          context.types().GetAs<SemIR::ClassType>(class_info.self_type_id);
+      ExportClassToCpp(context, class_type);
```
Comment (danakj) [inline toolchain/check/class.cpp]: ExportClassToCpp can fail if it diagnoses an error. One way I can see that happening right now is if the class is generic.

https://github.com/carbon-language/carbon-lang/blob/eabc7f78b22c4ff4638c5d76d27435aa166245f7/toolchain/check/cpp/export.cpp#L66-L69

Can we add a test where the Carbon subclass class is generic? Do we still crash there?

File: toolchain/lower/file_context.cpp [hunk 4]
```diff
@@ -772,8 +772,14 @@ auto FileContext::BuildVtable(const SemIR::Vtable& vtable,
     -> llvm::Constant* {
   const auto& class_info = sem_ir().classes().Get(vtable.class_id);
   if (!vtable.carbon_native_vtable) {
-    auto* cxx_record_decl = cast<clang::CXXRecordDecl>(
-        sem_ir().clang_decls().Lookup(class_info.latest_decl_id())->key.decl);
+    const auto* clang_decl =
+        sem_ir().clang_decls().Lookup(class_info.first_decl_id());
+    if (!clang_decl) {
+      clang_decl = sem_ir().clang_decls().Lookup(class_info.latest_decl_id());
+    }
```
Comment (danakj) [inline toolchain/lower/file_context.cpp]: Why do we try the first_decl_id first now, but fall back to the latest? The PR description says the same, that we do this, but doesn't explain either.

The export uses the first decl:
https://github.com/carbon-language/carbon-lang/blob/eabc7f78b22c4ff4638c5d76d27435aa166245f7/toolchain/check/cpp/export.cpp#L284

So when is the latest one helpful?

File: toolchain/check/class.cpp (hunk 2 above)
Comment (jaswanthsanjay88) [reply toolchain/check/class.cpp]: Applied the suggestion, thanks.

File: toolchain/check/class.cpp (hunk 3 above)
Comment (jaswanthsanjay88) [reply toolchain/check/class.cpp]: Removed the fallback. first_decl_id() is the canonical declaration ID used to key clang_decls() (both during export and import), so the fallback to latest_decl_id() was unnecessary.

File: toolchain/lower/file_context.cpp (hunk 4 above)
Comment (jaswanthsanjay88) [reply toolchain/lower/file_context.cpp]: Added GenericZ(T: type) extending Cpp.X to virtual_fn_carbon_only.carbon. The generic definition is exported as a ClassTemplateDecl, and CheckCompleteClassType completes cleanly. If a specific class fails to map to C++, ExportClassToCpp diagnoses a SemanticsTodo error, which halts compilation before lowering without crashing.

File: toolchain/check/class.cpp [hunk 5]
```diff
@@ -417,12 +417,9 @@ static auto CheckCompleteClassType(
     auto vptr_type_id = GetPointerType(context, SemIR::VtableType::TypeInstId);
     class_info.vtable_decl_id = AddInst<SemIR::VtableDecl>(
         context, node_id, {.type_id = vptr_type_id, .vtable_id = vtable_id});
-    if (!context.vtables().Get(vtable_id).carbon_native_vtable &&
-        class_info.self_type_id.has_value() &&
-        class_info.self_type_id != SemIR::ErrorInst::TypeId) {
-      auto class_type =
-          context.types().GetAs<SemIR::ClassType>(class_info.self_type_id);
-      ExportClassToCpp(context, class_type);
+    if (!context.vtables().Get(vtable_id).carbon_native_vtable) {
```
Comment (danakj) [inline toolchain/check/class.cpp]: One more suggestion. We just built this vtable_id above, but now we're doing a lookup for it again, which requires us to go through a value store. Could we have BuildVtable return the id _and_ the carbon_native_vtable bool (as a pair), and then we have all the information here directly instead of needing to use the vtables() store?

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon [hunk 6]
```diff
@@ -20,6 +20,11 @@ class Z {
   override fn f(unused self) {}
 }
 
+class GenericZ(T: type) {
+  extend base: Cpp.X;
+  override fn f(unused self) {}
+}
+
```
Comment (danakj) [inline toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]: Let's put this in a separate file split (the `// --- file name` you see in other test files.

And let's include a use of it there like we do for Z.

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon [hunk 7]
```diff
@@ -20,6 +20,11 @@ class Z {
...
 fn Run() {
```
Comment (danakj) [inline toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]: Tests dont need a Run() function, we generally use shorter names.
```suggestion
fn F() {
```

File: toolchain/check/class.cpp (hunk 5 above)
Comment (jaswanthsanjay88) [reply toolchain/check/class.cpp]: Updated BuildVtable to return std::pair<SemIR::VtableId, bool> and used the bool directly in CheckCompleteClassType without querying `vtables()`.

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon (hunk 6 above)
Comment (jaswanthsanjay88) [reply toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]: Split into `non_generic.carbon` and `generic.carbon`, added `var _: GenericZ(type) = {.base = Cpp.X.X()};` to use the generic class, and switched to `min_prelude/destroy.carbon` (since class definitions require `Core.Destroy`).

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon (hunk 7 above)
Comment (jaswanthsanjay88) [reply toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]: Renamed to `fn F()` across both splits.

**Review by danakj — APPROVED** [review]
LGTM

**Review by danakj — CHANGES_REQUESTED** [review]
Ah looks like there is a crash with the generic function maybe?

```
1.	Program arguments: --include-diagnostic-kind compile --phase=lower --exclude-dump-file-prefix=/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/install/core --dump-llvm-ir --target=x86_64-linux-gnu min_prelude/destroy.carbon min_prelude/parts/destroy.carbon non_generic.carbon generic.carbon --custom-core --exclude-dump-file-prefix=min_prelude/ --clang-arg=-fno-exceptions
2.	Test stdout: (none)
	Test stderr: (none)
3.	Carbon version: 0.0.0-0.dev+unknown
4.	Pending diagnostics:
5.	Filename: non_generic.carbon
................ #0 0x0000aae710eac224 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x763c224)
 #1 0x0000aae710ea99d8 llvm::sys::RunSignalHandlers() (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x76399d8)
 #2 0x0000aae710e577e8 CrashRecoverySignalHandler(int) CrashRecoveryContext.cpp:0:0
 #3 0x0000ff68fd0629d0 (linux-vdso.so.1+0x9d0)
 #4 0x0000aae70e196ddc clang::ASTContext::getASTRecordLayout(clang::RecordDecl const*) const (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x4926ddc)
 #5 0x0000aae70e26bb44 (anonymous namespace)::FinalOverriders::FinalOverriders(clang::CXXRecordDecl const*, clang::CharUnits, clang::CXXRecordDecl const*) VTableBuilder.cpp:0:0
 #6 0x0000aae70e264168 (anonymous namespace)::ItaniumVTableBuilder::ItaniumVTableBuilder(clang::ItaniumVTableContext&, clang::CXXRecordDecl const*, clang::CharUnits, bool, clang::CXXRecordDecl const*) VTableBuilder.cpp:0:0
 #7 0x0000aae70e263bc4 clang::ItaniumVTableContext::computeVTableRelatedInformation(clang::CXXRecordDecl const*) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x49f3bc4)
 #8 0x0000aae70c3bb1b0 (anonymous namespace)::ItaniumCXXABI::getAddrOfVTable(clang::CXXRecordDecl const*, clang::CharUnits) ItaniumCXXABI.cpp:0:0
 #9 0x0000aae70c3baaec (anonymous namespace)::ItaniumCXXABI::getVTableAddressPoint(clang::BaseSubobject, clang::CXXRecordDecl const*) ItaniumCXXABI.cpp:0:0
#10 0x0000aae70bd32f40 Carbon::Lower::FileContext::PrepareToLower() (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x24c2f40)
#11 0x0000aae70bd32334 Carbon::Lower::Context::GetFileContext(Carbon::SemIR::File const*, Carbon::SemIR::InstNamer const*) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x24c2334)
#12 0x0000aae70bd315d8 Carbon::Lower::LowerToLLVM(llvm::LLVMContext&, llvm::IntrusiveRefCntPtr<llvm::vfs::FileSystem>, Carbon::FixedSizeValueStore<Carbon::SemIR::CheckIRId, llvm::function_ref<Carbon::Parse::TreeAndSubtrees const& ()>, Carbon::Untagged> const&, Carbon::SemIR::File const&, int, Carbon::Lower::LowerToLLVMOptions const&) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x24c15d8)
#13 0x0000aae70bd1be54 void llvm::function_ref<void ()>::callback_fn<Carbon::CompilationUnit::RunLower()::$_0>(long) compile_driver.cpp:0:0
#14 0x0000aae70bd12254 Carbon::CompilationUnit::LogCall(llvm::StringLiteral, llvm::StringLiteral, llvm::function_ref<void ()>) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x24a2254)
#15 0x0000aae70bd167a0 Carbon::CompileDriver::Compile(Carbon::DriverEnv&) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x24a67a0)
#16 0x0000aae70bcc6a84 Carbon::CompileSubcommand::Run(Carbon::DriverEnv&) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x2456a84)
#17 0x0000aae70bcbf6dc Carbon::Driver::RunCommand(llvm::ArrayRef<llvm::StringRef>) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x244f6dc)
#18 0x0000aae70bbf6328 Carbon::Testing::(anonymous namespace)::ToolchainFileTest::Run(llvm::SmallVector<llvm::StringRef, 3u> const&, llvm::IntrusiveRefCntPtr<llvm::vfs::InMemoryFileSystem>&, _IO_FILE*, llvm::raw_pwrite_stream&, llvm::raw_pwrite_stream&) const file_test.cpp:0:0
#19 0x0000aae70bc30290 Carbon::Testing::RunTestFile(Carbon::Testing::FileTestBase const&, bool, Carbon::Testing::TestFile&) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x23c0290)
#20 0x0000aae70bc2d218 Carbon::Testing::RunSingleTestHelper(Carbon::Testing::FileTestInfo&, Carbon::Testing::FileTestBase&) file_test_base.cpp:0:0
#21 0x0000aae710e57570 llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x75e7570)
#22 0x0000aae70bc2c7ac llvm::detail::UniqueFunctionBase<void>::UniqueFunctionBase<Carbon::Testing::FileTestEventListener::OnTestProgramStart(testing::UnitTest const&)::$_0, Carbon::Testing::FileTestEventListener::OnTestProgramStart(testing::UnitTest const&)::$_0>(Carbon::Testing::FileTestEventListener::OnTestProgramStart(testing::UnitTest const&)::$_0, llvm::detail::UniqueFunctionBase<void>::CalledAs<Carbon::Testing::FileTestEventListener::OnTestProgramStart(testing::UnitTest const&)::$_0>)::'lambda'(llvm::detail::UniqueFunctionBase<void> const*)::__invoke(llvm::detail::UniqueFunctionBase<void> const*) file_test_base.cpp:0:0
#23 0x0000aae70bc2c3f4 std::__1::__deferred_assoc_state<void, std::__1::__async_func<llvm::unique_function<void ()>>>::__execute() file_test_base.cpp:0:0
#24 0x0000aae710ec5aac llvm::StdThreadPool::processTasks(llvm::ThreadPoolTaskGroup*) (/home/runner/.cache/bazel/_bazel_runner/8f839eaeb716f9d034eabdfa7ebecdb0/sandbox/linux-sandbox/4837/execroot/_main/bazel-out/aarch64-opt/bin/toolchain/testing/file_test.runfiles/_main/toolchain/testing/file_test+0x7655aac)
#25 0x0000aae710ec6c30 void* llvm::thread::ThreadProxy<std::__1::tuple<llvm::StdThreadPool::grow(int)::$_0>>(void*) ThreadPool.cpp:0:0
#26 0x0000ff68fce503c8 (/lib/aarch64-linux-gnu/libc.so.6+0x803c8)
#27 0x0000ff68fceb9edc (/lib/aarch64-linux-gnu/libc.so.6+0xe9edc)
```

**Comment by jaswanthsanjay88** [comment]
The crash was actually in `non_generic.carbon` during lowering's `GetAddrOfVTable`: because `Z` was never referenced from C++, Clang never triggered external completion, leaving `Z` as an incomplete `CXXRecordDecl` when `getASTRecordLayout` was called.

I've updated the check phase to set `complete_type_witness_id` and call `CompleteType` on the exported Clang declaration so its definition, base classes, and overrides are complete before lowering. Generic dynamic classes with virtual C++ bases now diagnose a `SemanticsTodo` since specific class interop is unsupported.

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon [hunk 8]
```diff
@@ -26,19 +23,3 @@ class Z {
...
-''';
-
-class GenericZ(T: type) {
-  extend base: Cpp.X;
-  override fn f(unused self) {}
-}
-
-fn F() {
-  var _: GenericZ(type) = {.base = Cpp.X.X()};
-}
```
Comment (danakj) [inline toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon:44]: How come the test is removed? It sounds like it should now be a fail_todo_ test and would print a diagnostic?

**Review by danakj — COMMENTED** [review]
I am not yet familiar with how we export stuff to C++ so I am unsure about that part. @zygoloid could you have a double check at the check changes?

File: toolchain/check/cpp/export.cpp [hunk 9]
```diff
@@ -292,6 +292,16 @@ auto ExportClassToCpp(Context& context, SemIR::ClassType class_type)
   return record_decl;
 }
 
+auto ExportAndCompleteClassToCpp(Context& context, SemIR::ClassType class_type)
+    -> clang::TagDecl* {
+  auto* tag_decl = ExportClassToCpp(context, class_type);
+  if (tag_decl && context.cpp_context() &&
+      context.ast_context().getExternalSource()) {
+    context.ast_context().getExternalSource()->CompleteType(tag_decl);
+  }
```
Comment (zygoloid) [inline toolchain/check/cpp/export.cpp:301]: There might be cases where we need to call `context.clang_sema().MarkVTableUsed(...)` here instead of just `CompleteType` -- `CompleteType` will inject the complete class definition into the Clang AST, but in principle there could be other things that we need to do in order for Clang to be ready to lower a vtable.

Can you give that a go and see if it also works? If it does, I think that's the better approach here.

**Review by zygoloid — COMMENTED** [review]
LG, thanks. One comment, but I think this is fine as-is if that doesn't work out.

It looks like we also have some more work to do to make Carbon generic classes deriving from C++ classes work properly (we need to inject the Carbon class into the clang AST during monomorphization so its vtable is available to lowering) -- but we can tackle that separately :)

File: toolchain/check/cpp/export.cpp (hunk 9 above)
Comment (jaswanthsanjay88) [reply toolchain/check/cpp/export.cpp:302]: Updated `ExportAndCompleteClassToCpp` to call `context.clang_sema().MarkVTableUsed(cxx_record_decl->getLocation(), cxx_record_decl);` instead of `CompleteType`.

File: toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon (hunk 8 above)
Comment (jaswanthsanjay88) [reply toolchain/lower/testdata/interop/cpp/class/virtual_fn_carbon_only.carbon]: In lowering tests (`toolchain/lower/testdata/`), all split files are compiled together in a single `carbon compile --phase=lower` invocation. If any split fails checking (such as with `SemanticsTodo`), the compiler stops before lowering, which prevents lowering from running for the non-generic split.

To test the diagnostic without blocking lowering coverage, I added the generic case as a `// --- fail_todo_generic.carbon` split in `toolchain/check/testdata/interop/cpp/class/export/override.carbon`.

File: toolchain/check/testdata/interop/cpp/class/export/override.carbon [hunk 10]
```diff
@@ -30,3 +30,22 @@ class B {
...
+library "[[@TEST_NAME]]";
+import Cpp;
+
+inline Cpp '''
+struct X {
+  virtual void f() {}
+};
+''';
+
+// CHECK:STDERR: fail_todo_generic.carbon:[[@LINE+4]]:1: error: semantics TODO: `generic class deriving from C++ virtual class` [SemanticsTodo]
```
Comment (danakj) [inline toolchain/check/testdata/interop/cpp/class/export/override.carbon:73]: Maybe add a TODO to test this in lowering once it passes check?

**Review by danakj — APPROVED** [review]
LGTM

File: toolchain/check/testdata/interop/cpp/class/export/override.carbon (hunk 10 above)
Comment (jaswanthsanjay88) [reply toolchain/check/testdata/interop/cpp/class/export/override.carbon:73]: Added the TODO comment above the split in `toolchain/check/testdata/interop/cpp/class/export/override.carbon`.

**Comment by jaswanthsanjay88** [comment]
Testing `MarkVTableUsed` here resulted in a crash because `MarkVTableUsed` begins with an `if (!Class->isDynamicClass())` check, which dereferences `DefinitionData`. Because `ExportClassToCpp` produces a forward declaration with external storage where `DefinitionData` is `nullptr` until completed, calling `MarkVTableUsed` directly crashes. I have restored calling `CompleteType` directly on the external AST source.

**Comment by danakj** [comment]
> Testing `MarkVTableUsed` here resulted in a crash because `MarkVTableUsed` begins with an `if (!Class->isDynamicClass())` check, which dereferences `DefinitionData`. Because `ExportClassToCpp` produces a forward declaration with external storage where `DefinitionData` is `nullptr` until completed, calling `MarkVTableUsed` directly crashes. I have restored calling `CompleteType` directly on the external AST source.

Ok thanks. Are you not able to run the tests locally? It would be worth figuring out how to do that so we can avoid uploading and reviewing changes that won't pass tests.

**Comment by jaswanthsanjay88** [comment]
I am developing on Windows and had been relying on CI because Carbon requires a Linux/macOS environment with Clang 21+, which wasn't provisioned locally on this machine yet. I am setting up the WSL environment with Clang 21 and Bazelisk now so all subsequent changes and file tests can be verified locally before pushing.

**Comment by jaswanthsanjay88** [comment]
I have now set up and verified local testing via WSL (Ubuntu-24 with Clang 21, libc++-21, and Bazelisk) so test suites can run locally prior to pushing.

The test failure in `toolchain/check/testdata/interop/cpp/class/import/override.carbon` was because `CompleteType` on Carbon classes with C++ virtual bases causes Clang to synthesize C++ virtual function thunks. For `OverloadX` and `OverloadY`, building thunks for the mismatched overloads (`f(Y)` and `f(X)`) triggers parameter conversion checking, emitting the conversion diagnostics during thunk synthesis.

The test expectations have been updated and verified locally with `file_test`.

File: toolchain/check/class.cpp
Comment (danakj) [inline toolchain/check/class.cpp]: Let's mention what the returned bool is here too?

**Review by danakj — APPROVED** [review]
Thanks, LGTM w/ one last comment thing

**Comment by jaswanthsanjay88** [comment]
Updated the comment to document the returned bool (`carbon_native_vtable`, false for foreign vtables from C++ bases). Also autoupdated `toolchain/lower/testdata/interop/cpp/class/virtual_fn.carbon` to match the adjusted thunk emission order from earlier class completion.

**Comment by danakj** [comment]
May I suggest you have your agent not continuously merge trunk into the PR? It drops the PR out of the merge queue.
